src/checkpoints/checkpoints.cpp

```diff
@@ -69,92 +71,211 @@ namespace cryptonote
+      checkpoint_t checkpoint = {};
+      checkpoint.type         = checkpoint_type::predefined_or_dns;
+      checkpoint.block_hash   = h;
+      m_points[height]        = checkpoint;
     }
-    m_points[height] = h;
+
     return true;
   }
   //---------------------------------------------------------------------------
-  bool checkpoints::is_in_checkpoint_zone(uint64_t height) const
+  static bool add_vote_if_unique(checkpoint_t &checkpoint, service_nodes::checkpoint_vote const &vote)
   {
-    return !m_points.empty() && (height <= (--m_points.end())->first);
+    CHECK_AND_ASSERT_MES(checkpoint.block_hash == vote.block_hash,
+                         false,
+                         "Developer error: Add vote if unique should only be called when the vote hash and checkpoint hash match");
+
+    // TODO(doyle): Factor this out, a similar function exists in service node deregister
+    CHECK_AND_ASSERT_MES(vote.voters_quorum_index < service_nodes::QUORUM_SIZE,
+                         false,
+                         "Vote is indexing out of bounds");
+
+    const auto compare_vote_and_voter_to_signature = [](service_nodes::checkpoint_vote const &a, service_nodes::voter_to_signature const &b) {
+      return a.voters_quorum_index < b.quorum_index;
+    };
+
+    auto signature_it = std::upper_bound(checkpoint.signatures.begin(), checkpoint.signatures.end(), vote, compare_vote_and_voter_to_signature);
+    if (signature_it == checkpoint.signatures.end() || signature_it->quorum_index != vote.voters_quorum_index)
+    {
+      service_nodes::voter_to_signature new_voter_to_signature = {};
+      new_voter_to_signature.quorum_index                      = vote.voters_quorum_index;
+      new_voter_to_signature.signature                         = vote.signature;
+      checkpoint.signatures.insert(signature_it, new_voter_to_signature);
+      return true;
+    }
+
+    return false;
   }
   //---------------------------------------------------------------------------
-  bool checkpoints::check_block(uint64_t height, const crypto::hash& h, bool& is_a_checkpoint) const
+  bool checkpoints::add_checkpoint_vote(service_nodes::checkpoint_vote const &vote)
   {
-    auto it = m_points.find(height);
-    is_a_checkpoint = it != m_points.end();
-    if(!is_a_checkpoint)
+   // TODO(doyle): Always accept votes. Later on, we should only accept votes up
+   // to a certain point, so that eventually once a certain number of blocks
+   // have transpired, we can go through all our "seen" votes and deregister
+   // anyone who has not participated in voting by that point.
+
+#if 0
+    uint64_t newest_checkpoint_height = get_max_height();
+    if (vote.block_height < newest_checkpoint_height)
       return true;
+#endif
 
-    if(it->second == h)
+    CRITICAL_REGION_LOCAL(m_lock);
+    std::vector<int> unique_vote_set(service_nodes::QUORUM_SIZE);
```

Comment:
Use `std::array`, since the size is known at compile time?

Reply:
Updated